### PR TITLE
Make changes to previous PR #465

### PR DESCRIPTION
**Pull request by felixfaisal**
This PR addresses changes requested in #465

### CODE REVIEW

**Comment by Gauthamastro** [comment]
@felixfaisal, change it back to your staging branch; it's tough to review with many unrelated changes.

File: pallets/polkadex-ido/src/lib.rs [hunk 1]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+    pub(super) type TokenWithdrawn<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        T::Hash,
+        bool,
+        ValueQuery,
+    >;
+    
+
+
```
Comment (serhii-temchenko) [inline pallets/polkadex-ido/src/lib.rs]: Remove extra spaces

File: pallets/polkadex-ido/src/lib.rs [hunk 2]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+        FundingRoundRegistered(T::Hash),
+        ParticipatedInRound(T::Hash, T::AccountId),
+        ParticipatedInRoundFailed(T::Hash, T::AccountId, sp_runtime::DispatchError),
+        RaiseClaimed(T::Hash, T::AccountId, BalanceOf<T>),
+        TokenClaimed(T::Hash, T::AccountId),
+        InvestmentWithdrawn(T::Hash, T::AccountId, BalanceOf<T>),
+        TokenWithdrawn(T::Hash, T::AccountId, BalanceOf<T>)
+    }
+
+
```
Comment (serhii-temchenko) [inline pallets/polkadex-ido/src/lib.rs]: Remove extra space.

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/polkadex-ido/src/lib.rs [hunk 3]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+        #[pallet::constant]
+        type TreasuryAccountId: Get<Self::AccountId>;
+        /// Balances Pallet
+        type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId> + fungible::Inspect<Self::AccountId>;
+        /// The basic amount of funds that must be reserved for an Polkadex.
+        #[pallet::constant]
+        type IDOPDXAmount: Get<BalanceOf<Self>>;
+        /// Maximum supply for IDO
+        #[pallet::constant]
+        type MaxSupply: Get<BalanceOf<Self>>;
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: MaxSupply of what?

File: pallets/polkadex-ido/src/lib.rs [hunk 4]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+        /// The generator used to supply randomness to IDO
+        type Randomness: Randomness<Self::Hash, Self::BlockNumber>;
+        /// Randomness Source for random participant seed
+        type RandomnessSource: Randomness<H256, Self::BlockNumber>;
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: RandomnessSource and Randomness are identical., remove one of it.

File: pallets/polkadex-ido/src/lib.rs [hunk 5]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+        /// Default investor locking period
+        #[pallet::constant]
+        type DefaultInvestorLockPeriod: Get<Self::BlockNumber>;
+        /// Minimum deposit to create PDEX account for round id
+        #[pallet::constant]
+        type ExistentialDeposit: Get<BalanceOf<Self>>;
+
+        /// One PDEX amount in u128
+        #[pallet::constant]
+        type OnePDEX : Get<u128>;
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: This should be defined as constant in polkadex-primitives and imported from there, I think it's already there as UNIT_BALANCE

File: pallets/polkadex-ido/src/lib.rs [hunk 6]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+
+        type AssetManager: Create<<Self as frame_system::Config>::AccountId>
+        + Mutate<<Self as frame_system::Config>::AccountId, Balance=u128, AssetId=u128>
+        + Inspect<<Self as frame_system::Config>::AccountId>
+        + Transfer<<Self as frame_system::Config>::AccountId>
+        + Unbalanced<<Self as frame_system::Config>::AccountId>;
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: Unbalanced? Why do we need it?

File: pallets/polkadex-ido/src/lib.rs [hunk 7]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+            vesting_per_block: BalanceOf<T>,
+            funding_period: T::BlockNumber,
+            min_allocation: BalanceOf<T>,
+            max_allocation: BalanceOf<T>,
+            token_a_price_per_token_b: BalanceOf<T>,
+        ) -> DispatchResult {
+            let team: T::AccountId = ensure_signed(origin)?;
+            //TODO check if funder have the token_a available and reserve them.
+            // CID len must be less than or equal to 100
+            ensure!(cid.len() <= 100, <Error<T>>::CidReachedMaxSize);
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: Configure 100 as a constant in a constants file.

File: pallets/polkadex-ido/src/lib.rs [hunk 8]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+            ensure!(min_allocation <= max_allocation, <Error<T>>::MinAllocationMustBeEqualOrLessThanMaxAllocation);
+            ensure!(vesting_per_block > Zero::zero(), <Error<T>>::VestingPerBlockMustGreaterThanZero);
+            let current_block_no = <frame_system::Pallet<T>>::block_number();
+            let start_block = current_block_no.clone().saturating_add(1_u128.saturated_into());
+            let close_round_block = current_block_no.saturating_add(funding_period);
+            let token_a_price_per_token_b_perquintill = Perbill::from_rational(token_a_price_per_token_b, 1_000_000_000_000_u128.saturated_into());
+            ensure!(!token_a_price_per_token_b_perquintill.is_zero(), <Error<T>>::PricePerTokenCantBeZero);
+            ensure!(start_block < close_round_block, <Error<T>>::StartBlockMustBeLessThanEndblock);
+            // TODO: Need to handle unwrap for overflows when it returns a none
+            let vesting_period: u32 = (amount.checked_div(&vesting_per_block).unwrap()).saturated_into();
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: Remove unwrap() before merging it

File: pallets/polkadex-ido/src/lib.rs [hunk 9]
```diff
@@ -125,1451 +63,628 @@ type BalanceOf<T> =
...
+            let mut funding_round = <InfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
+            let current_block_no = <frame_system::Pallet<T>>::block_number();
+            ensure!(current_block_no >= funding_round.start_block && current_block_no < funding_round.close_round_block, <Error<T>>::NotAllowed);
+            ensure!(Self::can_withdraw(funding_round.token_b,&investor_address, amount.saturated_into()).is_ok(), Error::<T>::BalanceInsufficientForInteresetedAmount);
+            let token_a_price_per_token_b_u128: u128 = funding_round.token_a_price_per_token_b.saturated_into();
+            let token_a_b_ratio = FixedU128::saturating_from_rational(1_000_000_000_000_u128, token_a_price_per_token_b_u128);
+            let amount_u128: u128 = amount.saturated_into();
+            let amount_in_fixed_point = FixedU128::from_inner(amount_u128);
+            // TODO: This unwrap needs to be handled properly incase there's a None value returned
+            let amount_in_token_a: BalanceOf<T> = token_a_b_ratio.checked_mul(&amount_in_fixed_point).unwrap().into_inner().saturated_into();
```
Comment (Gauthamastro) [inline pallets/polkadex-ido/src/lib.rs]: Remove all unwraps

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/polkadex-ido/src/lib.rs (hunk 1 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Done ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 2 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Done ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 3 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Previous implementation code, Removed ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 4 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Removed RandomnessSource ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 5 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Removed ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 6 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: We do not need it, Removed ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 7 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Done ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 9 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Done ✅

File: pallets/polkadex-ido/src/lib.rs (hunk 8 above)
Comment (felixfaisal) [reply pallets/polkadex-ido/src/lib.rs]: Done ✅

**Review by Gauthamastro — APPROVED** [review]

**Review by serhii-temchenko — APPROVED** [review]

**Pull request merged**